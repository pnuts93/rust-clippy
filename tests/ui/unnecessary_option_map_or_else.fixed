#![warn(clippy::unnecessary_option_map_or_else)]
#![allow(
    clippy::let_and_return,
    clippy::let_unit_value,
    clippy::unnecessary_lazy_evaluations,
    clippy::unnecessary_literal_unwrap
)]

fn main() {
    // Expected errors
    // Basic scenario
    let option = Some(());
    option.unwrap_or_else(|| ()); //~ ERROR: unused "map closure" when calling

    // Type ascription
    let option = Some(());
    option.unwrap_or_else(|| ()); //~ ERROR: unused "map closure" when calling

    // Auto-deref
    let string = String::new();
    let option = Some(&string);
    let _: &str = option.unwrap_or_else(|| &string); //~ ERROR: unused "map closure" when calling

    // Temporary variable
    let option = Some(());
    option.unwrap_or_else(|| ());

    // Correct usages
    let option = Some(());
    option.map_or_else(|| (), |_| ());

    let option = Some(());
    option.map_or_else(|| (), |_: ()| ());

    let string = String::new();
    let option = Some(&string);
    let _: &str = option.map_or_else(|| &string, |_| &string);

    let option = Some(());
    option.map_or_else(
        || (),
        |_| {
            let tmp = ();
            tmp
        },
    );
}
