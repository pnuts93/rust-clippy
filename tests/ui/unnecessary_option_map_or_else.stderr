error: unused "map closure" when calling `Option::map_or_else` value
  --> tests/ui/unnecessary_option_map_or_else.rs:13:5
   |
LL |     option.map_or_else(|| (), |x| x);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `unwrap_or_else`: `option.unwrap_or_else(|| ())`
   |
   = note: `-D clippy::unnecessary-option-map-or-else` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unnecessary_option_map_or_else)]`

error: unused "map closure" when calling `Option::map_or_else` value
  --> tests/ui/unnecessary_option_map_or_else.rs:17:5
   |
LL |     option.map_or_else(|| (), |x: ()| x);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `unwrap_or_else`: `option.unwrap_or_else(|| ())`

error: unused "map closure" when calling `Option::map_or_else` value
  --> tests/ui/unnecessary_option_map_or_else.rs:22:19
   |
LL |     let _: &str = option.map_or_else(|| &string, |x| x);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `unwrap_or_else`: `option.unwrap_or_else(|| &string)`

error: unused "map closure" when calling `Option::map_or_else` value
  --> tests/ui/unnecessary_option_map_or_else.rs:26:5
   |
LL | /     option.map_or_else(
LL | |
LL | |         || (),
LL | |         |x| {
...  |
LL | |         },
LL | |     );
   | |_____^ help: consider using `unwrap_or_else`: `option.unwrap_or_else(|| ())`

error: aborting due to 4 previous errors

